components/eamxx/src/physics/mam/impl/mam4_amicphys.cpp

```diff
@@ -1389,197 +1735,427 @@ void mam_amicphys_1gridcell(
+          qaer3[iaer][imode] = qsub3[la][jsub] * fcvt_aer(iaer);
+          qaer4[iaer][imode] = qaer3[iaer][imode];
+        }
+      }  // for iaer
+      qwtr3[imode] = qaerwatsub3[imode][jsub] * fcvt_wtr();
+      qwtr4[imode] = qwtr3[imode];
+    }  // for imode
+
+    Real qaercw2[num_aerosol_ids][num_modes] = {0};
+    Real qnumcw2[num_modes]                  = {0};
+    Real qaercw3[num_aerosol_ids][num_modes] = {0};
+    Real qnumcw3[num_modes]                  = {0};
+    Real qaercw4[num_aerosol_ids][num_modes] = {0};
+    Real qnumcw4[num_modes]                  = {0};
+
+    assign_2d_array(num_aerosol_ids, num_modes, 0.0,  // in
+                    &qaercw2[0][0]);                  // out
+    assign_2d_array(num_aerosol_ids, num_modes, 0.0,  // in
+                    &qaercw3[0][0]);                  // out
+    assign_2d_array(num_aerosol_ids, num_modes, 0.0,  // in
+                    &qaercw4[0][0]);                  // out
+
+    assign_1d_array(num_modes, 0.0,  // in
+                    qnumcw2);        // out
+    assign_1d_array(num_modes, 0.0,  // in
+                    qnumcw3);        // out
+    assign_1d_array(num_modes, 0.0,  // in
+                    qnumcw4);        // out
+
+    if(iscldy_subarea[jsub]) {
+      for(int imode = 0; imode < num_modes; ++imode) {
+        qnumcw2[imode] = 0;
+        qnumcw3[imode] = 0;
+        qnumcw4[imode] = 0;
+      }  // imode
+      for(int iaer = 0; iaer < num_aerosol_ids; ++iaer) {
+        for(int imode = 0; imode < num_modes; ++imode) {
+          qaercw2[iaer][imode] = 0;
+          qaercw3[iaer][imode] = 0;
+          qaercw4[iaer][imode] = 0;
+        }  // imode
+      }    // iaer
       // only do cloud-borne for cloudy
-      for (int n = 0; n < num_modes; ++n) {
-        qnumcw3[n] = qqcwsub3[n][jsub] * fcvt_num();
-        qnumcw4[n] = qnumcw3[n];
-        for (int iaer = 0; iaer < num_aerosol_ids; ++iaer) {
-          qaercw2[iaer][n] = qqcwsub2[n][jsub] * fcvt_aer(iaer);
-          qaercw3[iaer][n] = qqcwsub3[n][jsub] * fcvt_aer(iaer);
-          qaercw4[iaer][n] = qaercw3[iaer][n];
+      for(int imode = 0; imode < num_modes; ++imode) {
+        int ln         = lmap_numcw(imode);
+        qnumcw2[imode] = qqcwsub2[ln][jsub] * fcvt_num();
+        qnumcw3[imode] = qqcwsub3[ln][jsub] * fcvt_num();
+        qnumcw4[imode] = qnumcw3[imode];
+      }  // imode
+      for(int iaer = 0; iaer < num_aerosol_ids; ++iaer) {
+        for(int imode = 0; imode < num_modes; ++imode) {
+          int la = lmap_aer(iaer, imode);
+          if(la > 0) {
+            qaercw2[iaer][imode] = qqcwsub2[la][jsub] * fcvt_aer(iaer);
+            qaercw3[iaer][imode] = qqcwsub3[la][jsub] * fcvt_aer(iaer);
+            qaercw4[iaer][imode] = qaercw3[iaer][imode];
+          }  // la
+        }    // imode
+      }      // iaer
+    }        // iscldy_subarea
+
+    Real qgas_delaa[max_gas()][nqtendaa()]                       = {};
```

Comment:
For clarity: the `={}` part initializes all entries to 0. If you don't need to init to 0, consider removing that part.

Reply:
I know we can init arrays using `={}` to a specific value. I read somewhere that it is true for CPU codes but C++ doesn't guarantee it for codes running on GPUs. Do you know if all modern GPUs will init arrays to init values if we use `={}`?

Reply:
The standard GUARANTEEES that doing `Real blah[10] = {};` will cause ALL entries to be value-initialized. For builtin types (like double/float, which are what Real is), this means they will be init-ed to 0.

This is specific to the fact that the array entries are of a builtin type AND you want them init-ed to 0. If they were not built-in types OR you wanted a different value, it would be another story.